src/link/database.rs

```diff
@@ -106,31 +114,66 @@ pub fn links_metadata(schema: &Schema) -> Result<Option<LinksMetadata>, LinkErro
     }))
 }
 
-// Note: currently only recognizing @link, not @core. Doesn't feel worth bothering with @core at
-// this point, but the latter uses the "feature" arg instead of "url".
-fn parse_link_if_bootstrap_directive(schema: &Schema, directive: &Directive) -> bool {
-    if let Some(definition) = schema.directive_definitions.get(&directive.name) {
-        let locations = &definition.locations;
-        let is_correct_def = definition.repeatable
-            && locations.len() == 1
-            && locations[0] == DirectiveLocation::Schema;
-        let is_correct_def = is_correct_def
-            && definition.arguments.iter().any(|arg| {
-                arg.name == "as" && matches!(&*arg.ty, Type::Named(name) if name == "String")
-            });
-        let is_correct_def = is_correct_def && definition.arguments.iter().any(|arg| {
-            arg.name == "url" && {
-                // The "true" type of `url` in the @link spec is actually `String` (nullable), and this
+/// Returns true if the given definition matches the @link definition.
+///
+/// Either of these definitions are accepted:
+/// ```graphql
+/// directive @_ANY_NAME_(url: String!, as: String) repeatable on SCHEMA
+/// directive @_ANY_NAME_(url: String, as: String) repeatable on SCHEMA
+/// ```
+fn is_link_directive_definition(definition: &DirectiveDefinition) -> bool {
+    definition.repeatable
+        && definition.locations == [DirectiveLocation::Schema]
+        && definition.argument_by_name("url").is_some_and(|argument| {
+            // The "true" type of `url` in the @link spec is actually `String` (nullable), and this
+            // for future-proofing reasons (the idea was that we may introduce later other
+            // ways to identify specs that are not urls). But we allow the definition to
+            // have a non-nullable type both for convenience and because some early
+            // federation previews actually generated that.
+            *argument.ty == ty!(String!) || *argument.ty == ty!(String)
+        })
+        && definition
+            .argument_by_name("as")
+            .is_some_and(|argument| *argument.ty == ty!(String))
+}
+
+/// Returns true if the given definition matches the @core definition.
+///
+/// Either of these definitions are accepted:
+/// ```graphql
+/// directive @_ANY_NAME_(feature: String!, as: String) repeatable on SCHEMA
+/// directive @_ANY_NAME_(feature: String, as: String) repeatable on SCHEMA
+/// directive @_ANY_NAME_(feature: String!) repeatable on SCHEMA
+/// directive @_ANY_NAME_(feature: String) repeatable on SCHEMA
+/// ```
+fn is_core_directive_definition(definition: &DirectiveDefinition) -> bool {
+    // XXX(@goto-bus-stop): @core compatibility is primarily to support old tests--should be
+    // removed when those are updated.
+    definition.repeatable
+        && definition.locations == [DirectiveLocation::Schema]
+        && definition
+            .argument_by_name("feature")
```

Comment:
Shouldn’t this also check that `defintion` does not have unexpected arguments?

Reply:
I would agree, though the JS code doesn't appear to check for this, maybe for forward compatibility..? I think that seems risky since the interpretation would likely be wrong if there's an unsupported argument.

Reply:
we should keep it the same as the JS code